oss_lib/config.py

```diff
@@ -0,0 +1,235 @@
+    }
+    return variables
+
+
+def setup_config(config_path=None, default_config_path=None, defaults=None,
+                 log_config_path=None, debug=False,
+                 validation_schema=None):
+    global CONF, _CONF
+
+    setup_logging(log_config_path, debug=debug)
+
+    if defaults:
+        config = copy.deepcopy(defaults)
+    else:
+        config = {}
+
+    loaded_config = load_config_file(config_path, default_config_path)
+    if loaded_config:
+        merge_dicts(config, loaded_config)
+
+    if debug:
+        log_config(config, config_path, default_config_path)
+
+    validate_config(config, validation_schema)
+    _CONF = config
+    return CONF
+
+
+def setup_logging(config_path, debug=False):
+    if not config_path:
+        logging.basicConfig(format=LOG_FORMAT, datefmt=LOG_DATE_FORMAT)
+    else:
+        LOG.info("Logging configuration file is specified: %s",
+                 config_path)
+        logging.config.fileConfig(config_path)
+    if debug:
+        LOG.info("Setting up the DEBUG logging level.")
+        root_logger = logging.getLogger()
+        root_logger.setLevel(logging.DEBUG)
+
+
+def load_config_file(config_path, default_config_path=None):
+    config = None
+    if config_path:
+        LOG.info("Configuration file was specified: %s", config_path)
+        config = yaml_load(config_path)
+    elif default_config_path and os.path.exists(default_config_path):
+        LOG.info("Trying to read default configuration file: %s",
+                 default_config_path)
+        config = yaml_load(default_config_path)
+    return config
+
+
+def yaml_load(path):
+    try:
+        with open(path) as f:
+            return yaml.safe_load(f)
+    except IOError as exc:
+        LOG.error("Could not read content from %s: %s", path, exc)
+        raise
+    except yaml.YAMLError as exc:
+        LOG.error("Could not parse YAML-content from %s: %s", path, exc)
+        raise
+
+
+def merge_dicts(source, other):
+    for key, other_value in six.iteritems(other):
+        try:
+            value = source[key]
+        except KeyError:
+            source[key] = other_value
+            continue
+        if (isinstance(value, dict) and
+                isinstance(other_value, dict)):
+            merge_dicts(value, other_value)
+        else:
+            source[key] = other_value
+
+
+def log_config(config, config_path, default_config_path):
+    LOG.debug("Configuration files: config_path=%s, default_config_path=%s",
+              config_path, default_config_path)
+    LOG.debug("Content of the configuration:")
+    content = io.BytesIO()
+    yaml.dump(config, stream=content, default_flow_style=False)
+    content.seek(0)
+    for line in content:
+        LOG.debug(line)
+
+
+def validate_config(config, validation_schema=None):
+    schema = {
+        "type": "object",
+        "$schema": "http://json-schema.org/draft-04/schema",
+        "properties": {},
+        "additionalProperties": False,
```

Comment:
I don't like an idea to skip validation because it will just postpone problems on later steps, when we do not expect them. I would prefer to require a validation always.